src/Middleware/HeaderPropagation/src/HeaderPropagationMessageHandler.cs

```diff
@@ -0,0 +1,43 @@
+using System;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.HeaderPropagation
+{
+    public class HeaderPropagationMessageHandler : DelegatingHandler
+    {
+        private readonly HeaderPropagationState _state;
+        private readonly HeaderPropagationOptions _options;
+
+        public HeaderPropagationMessageHandler(IOptions<HeaderPropagationOptions> options, HeaderPropagationState state)
+        {
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            _options = options.Value;
+
+            _state = state ?? throw new ArgumentNullException(nameof(state));
+        }
+
+        protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request,
+            CancellationToken cancellationToken)
+        {
+            foreach (var header in _options.Headers)
+            {
+                if (_state.Headers.TryGetValue(header.OutputName, out var values) &&
+                    !StringValues.IsNullOrEmpty(values) &&
```

Comment:
@Tratcher - can you think of a case where a *present but empty* header is meaningful? I see the checks for null or empty and I wonder...

Reply:
The only case I've seen is the Host header. You're required to have one but it may be empty.

Reply:
OK, well that's one we deffo don't care about propagating.